File: Sources/Shared/Syntax/Parsers/YAMLFrontmatterParser.swift
// MARK: - YAMLFrontmatterParser
// Syntax parser for YAML frontmatter blocks (--- delimited).
// macOS 14+, Swift 5.10

import Foundation

/// Parses YAML frontmatter between `---` delimiters.
///
/// Emits ``SyntaxTokenKind/frontmatterDelimiter``,
/// ``SyntaxTokenKind/frontmatterKey``, and
/// ``SyntaxTokenKind/frontmatterValue`` tokens.
///
/// Not a standalone parser -- used as a sub-parser by ``CodeSpeakParser``
/// and any future composite parser that needs frontmatter support.
struct YAMLFrontmatterParser: SyntaxParsing, Sendable {

    let supportedExtensions: [String] = [] // Not standalone

    func parseLine(
        _ line: String,
        lineRange: NSRange,
        context: LineContext
    ) -> (tokens: [SyntaxToken], nextContext: LineContext) {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        var ctx = context

        // Frontmatter delimiter
        if trimmed == "---" {
            ctx.inFrontmatter.toggle()
            return (
                [SyntaxToken(kind: .frontmatterDelimiter, range: lineRange)],
                ctx
            )
        }

        guard ctx.inFrontmatter else {
            return ([], ctx)
        }

        // Key: value parsing
        let nsLine = line as NSString
        if let colonRange = line.range(of: ":") {
            let keyEnd = line.distance(from: line.startIndex, to: colonRange.lowerBound)
            let keyNSRange = NSRange(
                location: lineRange.location,
                length: min(keyEnd, nsLine.length)
            )
            let valueStart = line.distance(from: line.startIndex, to: colonRange.upperBound)
            let valueLen = max(0, nsLine.length - valueStart)
            let valueNSRange = NSRange(
                location: lineRange.location + valueStart,
                length: valueLen
            )

            var tokens: [SyntaxToken] = []
            if keyNSRange.length > 0 {
                tokens.append(SyntaxToken(kind: .frontmatterKey, range: keyNSRange))
            }
            if valueNSRange.length > 0 {
                tokens.append(SyntaxToken(kind: .frontmatterValue, range: valueNSRange))
            }
            return (tokens, ctx)
        }

        // Plain frontmatter line (array items, etc.)
        return (
            [SyntaxToken(kind: .frontmatterValue, range: lineRange)],
            ctx
        )
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/service/syntax/YAMLFrontmatterParser.kt
package studio.vibe.shared.service.syntax

import studio.vibe.shared.contract.LineContext
import studio.vibe.shared.contract.ParseLineResult
import studio.vibe.shared.contract.SyntaxParsing
import studio.vibe.shared.contract.SyntaxToken
import studio.vibe.shared.contract.SyntaxTokenKind

/**
 * Parses YAML frontmatter between `---` delimiters.
 *
 * Emits [SyntaxTokenKind.FRONTMATTER_DELIMITER], [SyntaxTokenKind.FRONTMATTER_KEY],
 * and [SyntaxTokenKind.FRONTMATTER_VALUE] tokens.
 *
 * Not a standalone parser — used as a sub-parser by [CodeSpeakSyntaxParser]
 * and any future composite parser that needs frontmatter support.
 * [supportedExtensions] is intentionally empty.
 */
class YAMLFrontmatterParser : SyntaxParsing {

    override val supportedExtensions: List<String> = emptyList()

    override fun parseLine(
        line: String,
        lineStartOffset: Int,
        lineEndOffset: Int,
        context: LineContext,
    ): ParseLineResult {
        val trimmed = line.trim()
        var ctx = context

        // Frontmatter delimiter `---`
        if (trimmed == "---") {
            ctx = ctx.copy(inFrontmatter = !ctx.inFrontmatter)
            return ParseLineResult(
                listOf(SyntaxToken(SyntaxTokenKind.FRONTMATTER_DELIMITER, lineStartOffset, lineEndOffset)),
                ctx,
            )
        }

        if (!ctx.inFrontmatter) {
            return ParseLineResult(emptyList(), ctx)
        }

        // Key: value parsing
        val colonIndex = line.indexOf(':')
        if (colonIndex >= 0) {
            val tokens = mutableListOf<SyntaxToken>()

            // Key token: from line start to colon (exclusive)
            val keyEnd = lineStartOffset + colonIndex
            if (keyEnd > lineStartOffset) {
                tokens.add(SyntaxToken(SyntaxTokenKind.FRONTMATTER_KEY, lineStartOffset, keyEnd))
            }

            // Value token: from after the colon to end of line
            val valueStart = lineStartOffset + colonIndex + 1
            if (valueStart < lineEndOffset) {
                tokens.add(SyntaxToken(SyntaxTokenKind.FRONTMATTER_VALUE, valueStart, lineEndOffset))
            }

            return ParseLineResult(tokens, ctx)
        }

        // Plain frontmatter line (array items, continuation lines, etc.)
        return ParseLineResult(
            listOf(SyntaxToken(SyntaxTokenKind.FRONTMATTER_VALUE, lineStartOffset, lineEndOffset)),
            ctx,
        )
    }
}
